tools/tool-env/profile-require.ts

```diff
@@ -0,0 +1,185 @@
+      return false;
+    }
+  
+    const ourSource = path.resolve(__dirname);
+    const required = path.resolve(path.dirname(this.filename), this.name);
+    if (ourSource.length > required.length) {
+      return false;
+    }
+    return required.substr(0, ourSource.length) === ourSource;
+  }
+
+  why(): string {
+    let walk: RequireInvocation = this;
+    let last: RequireInvocation | null = null;
+  
+    while (walk && ! walk.isOurCode()) {
+      last = walk;
+      walk = walk.parent;
+    }
+  
+    if (!walk) {
+      return "???";
+    }
+    if (last) {
+      return `${path.basename(walk.name)}:${path.basename(last.name)}`;
+    }
+    return path.basename(walk.name);
+  }
+}
+
+export function start() {
+  const moduleModule = require('module');
+  currentInvocation = new RequireInvocation('TOP');
+
+  const realLoader = moduleModule._load;
+  moduleModule._load = function (...args: [string, { filename?: string }]) {
+    const [id, { filename }] = args;
+    const inv = new RequireInvocation(id, filename);
+    const parent = currentInvocation;
+
+    currentInvocation.children.push(inv);
+    currentInvocation = inv;
+
+    try {
+      return realLoader.apply(this, args);
+    } finally {
+      inv.timeFinished = getNow();
+      currentInvocation = parent;
+    }
+  };
+}
+
+/**
+ * The discrepancy in computed times that we are willing to tolerate
+ * before deciding that the times don't add up.
+ */
+const TOLERANCE = 1/1000;
+
+export function printReport() {
+  currentInvocation.timeFinished = getNow();
+  computeTimes(currentInvocation);
+
+  const summary = summarize(currentInvocation, 0);
+  const times = Object.values(summary).sort((a, b) => 
+    b.time - a.time
+  );
+
+  let ourTotal = 0, otherTotal = 0;
+
+  times.forEach(item => {
+    let line = `${formatTime(item.time)} ${item.name}`;
+
+    if (!item.ours) {
+      line += ` [via ${Object.keys(item.via).join(', ')}]`;
+    }
+
+    console.log(line);
+
+    if (item.ours) {
+      ourTotal += item.time;
+    } else {
+      otherTotal += item.time;
+    }
+  })
+
+  const grandTotal = currentInvocation.totalTime || 0;
+  if (grandTotal - ourTotal - otherTotal > TOLERANCE) {
+    throw new Error("Times don't add up");
+  }
+  
+  console.log("TOTAL: ours " + formatTime(ourTotal) +
+              ", other " + formatTime(otherTotal) +
+              ", grand total " + formatTime(grandTotal));
+}
+
+function computeTimes(inv: RequireInvocation) {
+  inv.totalTime = inv.timeFinished! - inv.timeStarted;
+
+  let childTime = 0;
+  inv.children.forEach(child => {
+    computeTimes(child);
+    childTime += child.totalTime!;
+  });
+
+  if (inv.totalTime !== null) {
+    inv.selfTime = inv.totalTime - childTime;
+  }
+};
+
+type Summary = Record<string, {
+  name: string;
+  time: number;
+  ours: boolean;
+  via: Record<string, boolean>;
+}>;
+
+function summarize(inv: RequireInvocation, depth: number, summary: Summary = {}) {
```

Comment:
The functions don't have return types, which is mostly fine since they don't return anything. But this one does return something, so maybe we should annotate it. Or is it better just to let TS figure it out through static analysis?

Reply:
Although I do sometimes prefer to add an explicit return type (as I did in the class methods of `RequireInvocation`), personally I don't think it's the kind of thing that ought to be a strict rule.

For relatively simple functions that aren't exported from a module, if `TypeScript` is able to infer the type, then it's also very easy to see the inferred return type in VSCode, so there's no real loss of "documentation" information either way. In that case, adding the return type just puts a little bit of clutter in the source code that doesn't really add any additional value.

The main reason why I might want to add a return type signature is to be extra sure that the code inside really returns what I think it returns, even if I start refactoring it, but the stakes are quite low for a function like this (that just returns one of its parameters, which is typed) or the one below (which is a trivial one-liner that converts a timestamp number into a string), because if the return type changes then the usage (which is limited to this file) will also start to throw TypeScript errors if the types are no longer compatible.

Of course, I wouldn't mind adding a return type (it is a very easy change to make), but the point I'm making is that I don't think it should be a "you must always do this 100% of the time" type of rule. I am happy to be overridden by any member of the Meteor team who would like to lay down the law on style guidelines, though -- if they want us to start adding return types everywhere, then of course I will!

Reply:
I could see arguments both for and against annotating everything, but looking at the Apollo codebase, it looks like return types are frequently omitted, even from exported functions. We could go with that as guidance, absent any explicit rule for Meteor.